### ISSUE TITLE
File::Find re-entrancy

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4067]
Migrated from [rt.perl.org#7099](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7099) (status was 'resolved')

Searchable as RT7099$

**Comment by p5pRT** [comment]
### From nobull@mail.com




---
File&#8203;::Find used to work re\-entrantly\. Now it does not\. This is a pain
as using find\(\) re\-entrantly can be a very powerfull technique\.

Evidently someone thought it would be nice to use lexical variables\.
It \_would\_ be nice to use lexical variable\.  Unfortunately local\(\)
doesn't work on lexicals\.  Simply removing the necessary local\(\)s was
\_not\_ the right solution\!

Fix&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- /usr/lib/perl5/5.6.0/File/Find.pm	Thu Feb 15 13:41:45 2001
+++ perl/File/Find.pm	Fri Jun  8 15:13:13 2001
@@ -181,8 +181,10 @@
 
 require File::Basename;
 
-my %SLnkSeen;
-my ($wanted_callback, $avoid_nlink, $bydepth, $no_chdir, $follow,
+# Should be my() but local() currently refuses to operate on lexicals
+
+our %SLnkSeen;
+our($wanted_callback, $avoid_nlink, $bydepth, $no_chdir, $follow,
     $follow_skip, $full_check, $untaint, $untaint_skip, $untaint_pat);
 
 sub contract_name {
@@ -280,6 +282,10 @@
 
     my $cwd           = $wanted->{bydepth} ? Cwd::fastcwd() : Cwd::cwd();
     my $cwd_untainted = $cwd;
+
+    local ($wanted_callback, $avoid_nlink, $bydepth, $no_chdir, $follow,
+	   $follow_skip, $full_check, $untaint, $untaint_skip, $untaint_pat);
+
     $wanted_callback  = $wanted->{wanted};
     $bydepth          = $wanted->{bydepth};
     $no_chdir         = $wanted->{no_chdir};
@@ -291,7 +297,11 @@
     $untaint_skip     = $wanted->{untaint_skip};
 
     # for compatability reasons (find.pl, find2perl)
-    our ($topdir, $topdev, $topino, $topmode, $topnlink);
+    local our ($topdir, $topdev, $topino, $topmode, $topnlink);
+
+    local %SLnkSeen;
+
+    local ($dir, $name, $fullname, $prune);
 
     # a symbolic link to a directory doesn't increase the link count
     $avoid_nlink      = $follow || $File::Find::dont_use_nlink;
@@ -704,14 +714,12 @@
 sub find {
     my $wanted = shift;
     _find_opt(wrap_wanted($wanted), @_);
-    %SLnkSeen= ();  # free memory
 }
 
 sub finddepth {
     my $wanted = wrap_wanted(shift);
     $wanted->{bydepth} = 1;
     _find_opt($wanted, @_);
-    %SLnkSeen= ();  # free memory
 }
 
 # These are hard-coded for now, but may move to hint files.


-----------------------------------------------------------------
---

```
</details>
Flags:
&nbsp;    category=library
&nbsp;    severity=medium


---
Site configuration information for perl v5\.6\.0&#8203;:

Configured by bam at Thu Feb 15 14&#8203;:14&#8203;:58 GMT 2001\.

Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.2\.16\, archname=i586\-linux
&nbsp;    uname='linux wcl\-l 2\.2\.16 \#1 wed aug 2 20&#8203;:22&#8203;:26 gmt 2000 i586 unknown '
&nbsp;    config\_args='\-d'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-O2'\, gccversion=2\.95\.2 19991024 \(release\)
&nbsp;    cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.6\.0&#8203;:
&nbsp;    /home/bam/perl5
&nbsp;    /usr/lib/perl5/5\.6\.0/i586\-linux
&nbsp;    /usr/lib/perl5/5\.6\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.6\.0/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.005/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.005
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl v5\.6\.0&#8203;:
&nbsp;    HOME=/home/bam
&nbsp;    LANG=en\_US
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/bam/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/bin&#8203;:/usr/lib/java/bin&#8203;:/var/lib/dosemu&#8203;:/usr/games/bin&#8203;:/usr/games&#8203;:/opt/gnome/bin&#8203;:/opt/kde/bin&#8203;:\.
&nbsp;    PERL5LIB=/home/bam/perl5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @pmqs

>
> This is a bug report for perl from nobull@&#8203;mail\.com
> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> File&#8203;::Find used to work re\-entrantly\. Now it does not\. This is a pain
> as using find\(\) re\-entrantly can be a very powerfull technique\.
>
> Evidently someone thought it would be nice to use lexical variables\.
> It \_would\_ be nice to use lexical variable\.  Unfortunately local\(\)
> doesn't work on lexicals\.  Simply removing the necessary local\(\)s was
> \_not\_ the right solution\!

A patch to the test harness for File&#8203;::Find to test re\-entrancy would prevent
someone from reversing your patch in the future\. An explicit comment in the
stating why local is used would also help\.

This feature should be documented\.

Paul


\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
Do You Yahoo\!?
Get your free @&#8203;yahoo\.com address at http&#8203;://mail\.yahoo\.com

**Comment by p5pRT** [comment]
### From @jhi

Agreed\.

> stating why local is used would also help\.

Agreed\.

> This feature should be documented\.

Agreed\.

I now put available the most current versions of lib/File/Find\.pm and
t/lib/filefind\.t as http&#8203;://www\.chaos\.wustl\.edu/~jhi/find@&#8203;10525\.tgz
If I could get the patch reworked as described above against the
10525 version\, I would integrate it in for Perl 5\.7\.2\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Done \- I think \- I was rather confused by the test script\, it didn't
seem to work for testing File&#8203;::Find in isolation \(it cleared @&#8203;INC\) and
seemed to print the wrong number of tests\.  

Also I don't have access to a Mac to test it on\.

> > An explicit comment stating why local is used would also help\.
>
> Agreed\.

Agreed \- done\.

> > This feature should be documented\.
>
> Agreed\.

Disagree\.  This isn't a feature it's the absence of a limitiation
\(bug?\)\.  I tried to work it into the documentation I can't see any
place in the documentation that it would logically fit\.  IMNSHO\, if
one publishes an API that involves a function that takes a callback as
an argument then the function should be safe to use rentrantly from
within callback unless otherwise stated\.

> If I could get the patch reworked as described above against the
> 10525 version\, I would integrate it in for Perl 5\.7\.2\. 

http&#8203;://wcl\-l\.bham\.ac\.uk/pub/bam/patches/perl/find@&#8203;10525\.diff

The changes the the test script may need some checking\.

**Comment by p5pRT** [comment]
### From @andk

 >> > This feature should be documented\.
 >> 
 >> Agreed\.

&nbsp;  > Disagree\.  This isn't a feature it's the absence of a limitiation
&nbsp;  > \(bug?\)\.  I tried to work it into the documentation I can't see any
&nbsp;  > place in the documentation that it would logically fit\.  IMNSHO\, if
&nbsp;  > one publishes an API that involves a function that takes a callback as
&nbsp;  > an argument then the function should be safe to use rentrantly from
&nbsp;  > within callback unless otherwise stated\.

Please consider that old perl fellows already "know" that File&#8203;::Find
isn't reentrant\. So it may indeed be worth a mention in the manpage
that this bug is fixed\. Maybe under a header BUGS or HISTORY and maybe
with a wording like

&nbsp;    Up to version X\.XX\, File&#8203;::Find produced incorrect results if
&nbsp;    called recursively\. During the development of perl 5\.8 this bug
&nbsp;    was fixed\. The first fixed version of File&#8203;::Find was X\.XX\.

Just my 2\,

**Comment by p5pRT** [comment]
### From @jhi

Thanks\, applied\.  I think I will take up the suggestion of Andreas\, too\.

> The changes the the test script may need some checking\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I concur\. Of course for completeness you should also mention the first
version where the bug was introduced since even older Perl fellows
already "know" that File&#8203;::Find was reentrant upto Perl 5\.5\.

**Comment by p5pRT** [comment]
### From @jhi

Well\, Perlarchaeologists welcome\, though pointing to a $VERSION will
then become difficult because F&#8203;::F didn't have one back then\.\.\.
one would need to figure out the first official version of Perl where
F&#8203;::F become broken\.

**Comment by p5pRT** [comment]
### From @andk

 >> >   Up to version X\.XX\, File&#8203;::Find produced incorrect results if
 >> >   called recursively\. During the development of perl 5\.8 this bug
 >> >   was fixed\. The first fixed version of File&#8203;::Find was X\.XX\.
 >> 
 >> I concur\. Of course for completeness you should also mention the first
 >> version where the bug was introduced since even older Perl fellows
 >> already "know" that File&#8203;::Find was reentrant upto Perl 5\.5\.

&nbsp;  > Well\, Perlarchaeologists welcome\, though pointing to a $VERSION will
&nbsp;  > then become difficult because F&#8203;::F didn't have one back then\.\.\.
&nbsp;  > one would need to figure out the first official version of Perl where
&nbsp;  > F&#8203;::F become broken\.

I'm too busy to start archaeological digging right now but it's easy
to safe the world by simply stating that from version 1\.01 File&#8203;::Find
was reentrant \(As Jarkko did already\, Thanks\!\)\. The fact that some
previous versions were also reentrant is irrelevant as there has never
been a test for that\.
